features/include/pcl/features/gasd.h

```diff
@@ -0,0 +1,370 @@
+       * \param[in] hists shape histograms
+       * \param[out] output output descriptor point cloud
+       * \param[in,out] pos current position of output descriptor point cloud
+       */
+      void
+      copyShapeHistogramsToOutput (const size_t grid_size,
+                                   const size_t hists_size,
+                                   const std::vector<Eigen::VectorXf> &hists,
+                                   PointCloudOut &output,
+                                   size_t &pos);
+  };
+
+  /** \brief GASDColorEstimation estimates the Globally Aligned Spatial Distribution (GASD) descriptor for a given
+   * point cloud dataset given XYZ and RGB data.
+   *
+   * The suggested PointOutT is pcl::GASDSignature984.
+   *
+   * \note If you use this code in any academic work, please cite:
+   *
+   *   - J. Lima, V. Teichrieb.
+   *     An Efficient Global Point Cloud Descriptor for Object Recognition and Pose Estimation.
+   *     In Proceedings of the 29th SIBGRAPI - Conference on Graphics, Patterns and Images,
+   *     Sao Jose dos Campos, Brazil, October 4-7 2016.
+   *
+   * \author Joao Paulo Lima
+   *
+   * Voxar Labs, Centro de Informatica, Universidade Federal de Pernambuco, Brazil
+   *
+   * Departamento de Estatistica e Informatica, Universidade Federal Rural de Pernambuco, Brazil
+   *
+   * \ingroup features
+   */
+  template <typename PointInT, typename PointOutT = GASDSignature984>
+  class GASDColorEstimation : public GASDEstimation<PointInT, PointOutT>
+  {
+    public:
+      using typename Feature<PointInT, PointOutT>::PointCloudOut;
+      typedef boost::shared_ptr<GASDColorEstimation<PointInT, PointOutT> > Ptr;
+      typedef boost::shared_ptr<const GASDColorEstimation<PointInT, PointOutT> > ConstPtr;
+
+      /** \brief Constructor.
+       * \param[in] view_direction view direction
+       * \param[in] shape_half_grid_size shape half grid size
+       * \param[in] shape_hists_size shape histograms size
+       * \param[in] color_half_grid_size color half grid size
+       * \param[in] color_hists_size color histograms size
+       * \param[in] shape_interp shape histograms interpolation method
+       * \param[in] color_interp color histograms interpolation method
+       */
+      GASDColorEstimation (const Eigen::Vector3f &view_direction = Eigen::Vector3f (0.0f, 0.0f, 1.0f),
+                           const size_t shape_half_grid_size = 3,
+                           const size_t shape_hists_size = 1,
+                           const size_t color_half_grid_size = 2,
+                           const size_t color_hists_size = 12,
+                           const typename GASDEstimation<PointInT, PointOutT>::HistogramInterpolationMethod shape_interp = GASDEstimation<PointInT, PointOutT>::INTERP_NONE,
+                           const typename GASDEstimation<PointInT, PointOutT>::HistogramInterpolationMethod color_interp = GASDEstimation<PointInT, PointOutT>::INTERP_NONE) :
```

Comment:
This modification caused some compilation errors in the unit test. How should I change this?

Reply:
You should remove `HistogramInterpolationMethod::` because you are using old-style enum, which implies that it's items belong to surrounding namespace.

Reply:
I was not entirely correct here. As I wrote, enum items belong to the surrounding namespace. In this case the surrounding namespace of the `HistogramInterpolationMethod` enum is the class where it was defined, i.e. `GASDEstimation<PointInT, PointOutT>`. This means that you need to write

```cpp
const HistogramInterpolationMethod shape_interp = GASDEstimation<PointInT, PointOutT>::INTERP_NONE
```

Now this is counter-intuitive, especially for the users of the class, because they will also prefix enum value with the base class name! This is how the initialization of color GASD will look like:

```cpp
GASDColorEstimation<PointXYZRGB, GASDSignature984> gasd(a, b, c, d, e, GASDEstimation<PointXYZRGB, GASDSignature984>::INTERP_TRILINEAR);
```

So I would propose to shift the enum out of the class.